### ISSUE TITLE
HL Version specification

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#8649]
When compiling unit tests with `-D hl-ver=1.11`, I get the following:

```
unit.spec.TestString
  test: FAILURE .......................................................................................................................................................FFFF.
    line: 177, expected true
    line: 178, expected true
    line: 179, expected false
    line: 180, expected false
unit.spec.TestUnicodeString
  test: FAILURE .............................................................................................FFFF..............
    line: 134, expected true
    line: 135, expected true
    line: 136, expected true
    line: 137, expected true
```

I works correctly with `-D hl-ver=1.11.0` but I think it should work with both.

**Comment by Simn** [comment]
We discussed this before in #8297 and agreed to stick to the semver specification for this.

**Comment by ncannasse** [comment]
I'm not talking about `#if version()` here. hl_ver is meant to select the API level, it should not include minor versions since minor versions does not modify the API.

**Comment by ncannasse** [comment]
Also, in all cases we should validate the input value.

**Comment by nadako** [comment]
(I think it's called "patch versions")

**Comment by Gama11** [comment]
@ncannasse Well, it might be related if you want to be able to use `hl_ver` in `version()` checks. Or do you mean that `-D hl_ver=1.11` on the command line should just be turned into `1.11.0` behind the scenes?

**Comment by ncannasse** [comment]
I keep thinking that for both version() and other checks, 1.11 should be a synonym of 1.11.0, because it actually is, although semver fundamentalists doesn't seem to agree ;-)

**Comment by ncannasse** [comment]
Related discussion in #8650

**Comment by RealyUniqueName** [comment]
If that is allowed then 
```
#if (hl_ver <= version("1.11"))
```
suddenly stops working after "1.11.1" release, which is confusing.

**Comment by ncannasse** [comment]
@RealyUniqueName true, but I don't see how using 1.11.0 will fix something.

OTOH, we could argue that `version("1.11")` means actually `1.11.*` so both `1.11.0` and `11.11.1` are `<=` to `version("1.11")`

**Comment by RealyUniqueName** [comment]
> I don't see how using 1.11.0 will fix something

It's explicit, thus less confusing.

> OTOH,  we could argue that version("1.11") means actually 1.11.* so both 1.11.0 and 11.11.1 are <= to version("1.11")

that was discussed in #8297

**Comment by ncannasse** [comment]
I keep thinking that the outcome of #8297 was not a good idea.
Most of the version checks are wrt major or minor, it is very rare to have to version against patch versions. Yet we made patch version mandatory, although 0 about 99% of the time, with issues such as the one I'm writing here (invalid user input)

**Comment by Simn** [comment]
I'm closing this again. Please just add this `.0` to the version where necessary. It's not worth deviating from an established standard just so that can be omitted.

**Comment by ncannasse** [comment]
In that case d41949288 needs to be merged into 4.0 branch

**Comment by RealyUniqueName** [comment]
Merged
